File: src/01-B-table_insee.R
#Title: Table de Passage PREMIUM
#Script: France

# Setup ----
library(here)
library(arrow)
library(tidyverse)
library(data.table)
library(readxl)

path_insee_tables <- "data/external/insee-table_de_passage/"

# Arrondissement municipal to commune level ----
ARR2COM <- read_excel(paste0(path_insee_tables, "table-appartenance-geo-communes-24.xlsx"), 
                                   sheet = "ARM", skip = 5) |> 
  select(code_com = COM, code_arr = CODGEO)


# Yearly table de passage ----
TABLE_PASSAGE_DF <- read_excel(paste0(path_insee_tables, "table_passage_annuelle_2024.xlsx"), 
                               sheet = "COM", skip = 5) |>
  select(starts_with("CODGEO_201"), starts_with("CODGEO_202"))

names(TABLE_PASSAGE_DF) <- names(TABLE_PASSAGE_DF) |> 
  str_replace("CODGEO_20", "code_insee") |>
  tolower() 

APP_COM24 <- read_excel(paste0(path_insee_tables, "table-appartenance-geo-communes-24.xlsx"), 
                        sheet = "COM", skip = 5) |> 
  select(code_insee24 = CODGEO, arr24 = ARR, bv2022 = BV2022)

# Yearly passage with Arrondissement ---- 
TABLE_PASSAGE_DF <- TABLE_PASSAGE_DF |>
  left_join(APP_COM24, by="code_insee24") |>
  select(code_insee24, arr24, bv2022, everything()) |> 
  filter(!is.na(arr24)&!is.na(bv2022)) |>
  unique()

# Export ----
write_feather(as.data.frame(TABLE_PASSAGE_DF), "results_building/t_passage.feather", compression = "zstd")
write_feather(as.data.frame(ARR2COM), "results_building/arr2com.feather", compression = "zstd")

File: src/01-B-table_insee.py
# ---- Import Libraries ----
import pandas as pd
from pathlib import Path

# Title: PREMIUM Passage Table
# Script : France

# Setup ----
path_insee_tables = Path("data/external/insee-table_de_passage/")
output_dir = Path("results_building")
output_dir.mkdir(parents=True, exist_ok=True)

# Arrondissement municipal to commune level ----
arr2com = (
    pd.read_excel(
        path_insee_tables / "table-appartenance-geo-communes-24.xlsx",
        sheet_name="ARM",
        skiprows=5,
        dtype={"COM": str, "CODGEO": str}
    )
    .loc[:, ["COM", "CODGEO"]]
    .rename(columns={"COM": "code_com", "CODGEO": "code_arr"})
)

# Yearly table de passage ----
table_passage_df = (
    pd.read_excel(
        path_insee_tables / "table_passage_annuelle_2024.xlsx",
        sheet_name="COM",
        skiprows=5
    )
)

# Select columns starting with CODGEO_201 or CODGEO_202
cols_to_keep = [c for c in table_passage_df.columns if c.startswith("CODGEO_201") or c.startswith("CODGEO_202")]
table_passage_df = table_passage_df.loc[:, cols_to_keep]

# Clean column names
table_passage_df.columns = [
    col.replace("CODGEO_20", "code_insee").lower()
    for col in table_passage_df.columns
]

# Commune membership for 2024
app_com24 = (
    pd.read_excel(
        path_insee_tables / "table-appartenance-geo-communes-24.xlsx",
        sheet_name="COM",
        skiprows=5
    )
    .loc[:, ["CODGEO", "ARR", "BV2022"]]
    .rename(columns={
        "CODGEO": "code_insee24",
        "ARR": "arr24",
        "BV2022": "bv2022"
    })
)

# Yearly passage with Arrondissement ----
# prepare for export (Merge, filter, reorder, and deduplicate)
merged = (
    table_passage_df
    .merge(app_com24, left_on="code_insee24", right_on="code_insee24", how="left")
    .assign(
        code_insee24=lambda df: df["code_insee24"],
        arr24=lambda df: df["arr24"],
        bv2022=lambda df: df["bv2022"]
    )
    .loc[:, ["code_insee24", "arr24", "bv2022"] + 
         [c for c in table_passage_df.columns if c != "code_insee24"]]
    .dropna(subset=["arr24", "bv2022"])
    .drop_duplicates()
)

# Export ----
merged.reset_index(drop=True).to_feather(
    output_dir / "t_passage.feather", compression="zstd"
)
arr2com.to_feather(
    output_dir / "arr2com.feather", compression="zstd"
)
